sashimi-webapp/src/database/data-modifier/dataAdd.js

```diff
@@ -1,26 +1,82 @@
-import exceptions from '../exceptions';
+import exceptions from 'src/database/exceptions';
+import constants from 'src/database/constants';
+import SqlCommands from 'src/database/sql-related/sqlCommands';
+import DateTime from 'src/database/generated-data/dateTime';
+import SqlArray from 'src/database/generated-data/sqlArray';
+import StringManipulator from 'src/database/stringManipulation';
 
-import constants from '../constants';
+const sqlCommands = new SqlCommands();
+const dateTime = new DateTime();
+const alasqlArray = new SqlArray();
+const stringManipulator = new StringManipulator();
 
-import SqlCommands from '../sql-related/sqlCommands';
+function isLowerUpperBoundWithinRange(lowerBound, upperBound, givenArray) {
+  return lowerBound < upperBound
+    && lowerBound >= 0
+    && upperBound < givenArray.length;
+}
 
-import DateTime from '../generated-data/dateTime';
+function exchange(givenArray, currIndex, thenIndex) {
+  const tempObject = givenArray[currIndex];
+  givenArray[currIndex] = givenArray[thenIndex];
+  givenArray[thenIndex] = tempObject;
+}
 
-import SqlArray from '../generated-data/sqlArray';
+function bubbleUp(givenArray, lowerBound, upperBound) {
+  if (isLowerUpperBoundWithinRange(lowerBound, upperBound, givenArray)) {
+    let prev = lowerBound;
+    for (let index = lowerBound + 1; index <= upperBound; index+=1) {
+      exchange(givenArray, prev, index);
+      prev = index;
+    }
+  }
+}
 
-const sqlCommands = new SqlCommands();
-const dateTime = new DateTime();
-const alasqlArray = new SqlArray();
+function generateUniqueNewFileName(queryFiles, defaultFileName) {
+  let newFileName = defaultFileName;
+  let duplicateCount = 0;
+  for (let fileIndex = 0; fileIndex < queryFiles.length; fileIndex+=1) {
+    const fileObject = queryFiles[fileIndex];
+    if (fileObject[constants.HEADER_FILE_MANAGER_FILE_NAME] === newFileName) {
+      duplicateCount += 1;
+      // generate new unique file name
+      newFileName = stringManipulator.stringConcat(
+      constants.DEFAULT_FILE_NAME_OMIT_FILE_TYPE, duplicateCount, '.md');
+    } else if (fileIndex < queryFiles.length) {
+      bubbleUp(queryFiles, fileIndex, queryFiles.length -1);
+      fileIndex -= 1;
+    }
+  }
+  return newFileName;
+}
```

Comment:
There seems to exist cases there this function will run into infinite loop.

Case 1: Folder contains `newFile1`, `newFile2` but not `newFile`
```javascript
const queryFiles = [
  { 'file_name': 'newFile1.md' },
  { 'file_name': 'newFile2.md' },
  { 'file_name': 'newFile3.md' }
];
```
When the `newFile` is absent, the code will enter this condition: `} else if (fileIndex < queryFiles.length) {` and iterate from the previous index `fileIndex -= 1;` and restart the loop.
But `newFile` is still always absent, making it stuck in the loop.


Case 2: Similar consequence when there is gap in-between the list
```javascript
const queryFiles = [
  { 'file_name': 'newFile.md' },
  { 'file_name': 'newFile2.md' },
  { 'file_name': 'newFile3.md' }
];
```
---

Here is the test script I ran for this `generateUniqueNewFolderName`
https://gist.github.com/amoshydra/4e7c85baeced40db260b89503ef3e674